### DISCUSSION TITLE
tailwind 4 `oklch` not compality with old browser

### DISCUSSION THREAD

**Discussion by jesse996** [tailwindlabs/tailwindcss#15356] [Help] [answered]
tailwind 4 `okhlc` not compality with old browser,how to config to use old `rbg` rather than `oklch`

**Comment by adamwathan** [comment, marked as answer]
Hey! Tailwind 4 uses a lot of other features that aren’t compatible with older browsers so just changing this wouldn’t be enough to fix it. For v4 we’re targeting Safari 16.4 and above and recent versions of Chrome/Firefox. For older browsers you should stick with v3 for now 👍🏻

**Reply by jesse996** [reply]
ok i see,thank you for your answer

**Reply by tobiaspickel** [reply]
Hey, 
we do use tailwind 3 in a huge project that needs to support very old browsers.
As stick to v3 for now most likely means stick to v3 for ever. 
What is the plan to maintain v3 for projects that are stuck now?  
Or are there any plans to improve the v4 compatibility for legacy projects?
I cannot imagine that we are the only ones facing this issue 😅.

**Reply by sapiyans** [reply]
I created a site on the Tailwind V4 and was totally broken transpilation it didn't work, so I created a plugin using V3 so it is not completely broken in the old browser. See before and after: https://www.npmjs.com/package/vite-plugin-tailwind-legacy?activetab=readme

**Comment by brennanangel** [comment]
[This blog post says Lightning CSS would transpile it in 4.0](https://tailwindcss.com/blog/tailwindcss-v4-alpha). 

> "Syntax transforms — modern CSS features like oklch() colors and media query ranges are transpired to syntax with better browser support."

@adamwathan Is this [Lightning CSS transpilation](https://lightningcss.dev/transpilation.html#lab-colors) coming, or do you have a suggestion on how to add these as a fallback for those v4.0 devs who need to support older browsers?

**Reply by adamwathan** [reply]
Tailwind CSS v4.0 depends on other modern CSS features like [`@property`](https://caniuse.com/?search=%40property) and [`color-mix()`](https://caniuse.com/?search=color-mix), both of which only work in browsers that also support OKLCH, so our browser targets for transpilation do not transpile OKLCH.

Even if we did transpile OKLCH back to another format, you still couldn't use v4 in those browsers because those browsers wouldn't support `@property`. That feature specifically is very critical to how v4 works internally, so it's not possible to use v4 in older browsers that don't support it.

**Reply by davecarlson** [reply]
@adamwathan - It feels this wasn't signposted enough in the release notes. We've got a load of new bug reports of invisbile buttons on *not that old* ios devices ( due to the lack of loklch support ), so we're now going to hwave to go through and 'downgrade' to 3 until a proper solution is found

**Reply by alexanderbuhler** [reply]
@davecarlson Gonna leave that here for you, if you dare to tinker: https://gist.github.com/alexanderbuhler/2386befd7b6b3be3695667cb5cb5e709

**Comment by alexanderbuhler** [comment]
Here's a starting point that I got my project running with. Not perfect, but good enough. If you find something broken, let me know and I might have another look.

https://gist.github.com/alexanderbuhler/2386befd7b6b3be3695667cb5cb5e709

**Comment by zounar** [comment]
In my case, I have to support those older browsers, but I don't want to stick with Tailwind 3 just because of it. By changing the colors to `rgb()` and opting out from the `@layer` CSS rule, I was able to partially add support for older browsers. It's not perfect, but decent enough.

Check out [this repo](https://github.com/zounar/tailwind-4-rgb-palette) where I document those changes.

**Comment by Jaystar-Bee** [comment]
Due to this issue, html2pdf throws error `Error: Attempting to parse an unsupported color function "oklab"`. Seems like this compatibility is going to cause major issues

**Comment by nes2024** [comment]
According to their docs:
[Browser support](https://tailwindcss.com/docs/compatibility#browser-support)
Tailwind CSS v4.0 is designed for and tested on modern browsers, and the core functionality of the framework specifically depends on these browser versions:

Chrome 111 (released March 2023)
Safari 16.4 (released March 2023)
Firefox 128 (released July 2024)

How can you design a product that is incompatible with many common devices! Wasted months and months of development time! Tailwind developers, you bunch of w@nkers!

**Comment by sapiyans** [comment]
I created a site on the ```Tailwind V4``` and was totally broken transpilation it didn't work, so I created a plugin using V3 so it is not completely broken in the old browser. See before and after: https://www.npmjs.com/package/vite-plugin-tailwind-legacy?activetab=readme
